src/Stratis.Bitcoin.Features.SmartContracts/PoA/SmartContractPoARuleRegistration.cs

```diff
@@ -51,21 +53,25 @@ public class SmartContractPoARuleRegistration : IRuleRegistration
             this.fullTxValidationRules = fullTxValidationRules;
         }
 
-        public virtual void RegisterRules(IConsensus consensus)
+        public virtual void RegisterRules(IServiceCollection services)
         {
-            this.baseRuleRegistration.RegisterRules(consensus);
+            // TODO: this is not needed anymore as the default rules are registered in network
+            new PoAConsensusRulesRegistration().RegisterRules(services);
 
             // Add SC-Specific partial rules
             var txValidationRules = new List<IContractTransactionPartialValidationRule>(this.partialTxValidationRules)
             {
                 new SmartContractFormatLogic()
             };
-            consensus.PartialValidationRules.Add(new AllowedScriptTypeRule(this.network));
-            consensus.PartialValidationRules.Add(new ContractTransactionPartialValidationRule(this.callDataSerializer, txValidationRules));
+            services.AddSingleton(typeof(IPartialValidationConsensusRule), typeof(AllowedScriptTypeRule));
+            //consensus.ConsensusRules.PartialValidationRules.Add(typeof(AllowedScriptTypeRule));
 
-            consensus.FullValidationRules.Add(new ContractTransactionFullValidationRule(this.callDataSerializer, this.fullTxValidationRules));
+            services.AddSingleton(typeof(IPartialValidationConsensusRule), typeof(ContractTransactionPartialValidationRule));
+            //consensus.ConsensusRules.PartialValidationRules.Add((Type) new ContractTransactionPartialValidationRule(this.callDataSerializer, txValidationRules));
```

Comment:
I think you have changed the logic here. There is more than one partial validation rule, and it seems that not all get applied within `ContractTransactionPartialValidationRule` (perhaps it is intentional). So by relying on injection here you are potentially causing additional rules to be injected for `txValidationRules`.

Reply:
yes the smart contract part is still work in progress, the main part which is the RulesContainer is in place though.
I need to find a way to allow registering rule dependencies and not break current SC hacks.

Reply:
It turns out that the only other partial validation rule is not actually injected by any of the current daemons. So doing this will be ok, I may change the mempool partial validation to rely purely on injection too.